File: backend/inflection_lookup.py
#!/usr/bin/env python3
"""Pure-local grammatical analysis of an exact inflected word form —
part of speech plus full inflection (person / number / gender / tense /
mood). Reads `data/inflection/<lang>.jsonl` (built once by
`data_builder/build_inflections.py` from the English-Wiktionary Kaikki
dump, filtered to the language's wordlist), lazily per language and
cached for the process's lifetime — no network, no per-request I/O after
the first lookup for a language. The file is plain uncompressed
JSON-lines so it can be inspected by hand:

    {"form": "humera", "analyses": [
        {"pos": "verb", "tags": "third-person singular future", "lemma": "humer"}]}

`describe_form(word, language)` returns `[(pos_code, description), ...]`,
e.g. `[("verb", 'verb, third-person singular future (of "humer")')]`,
the exact shape `backend/clues.py`'s `_build_pos_block` consumes. `[]`
means the form isn't in the table (a lemma with no `form-of` entry, or a
form Wiktionary doesn't document) — the caller then falls back to the
Hunspell-stem + gloss-dictionary part-of-speech line.
"""
import json
from pathlib import Path

_DIR = Path(__file__).resolve().parent.parent / "data" / "inflection"

# Readable label for the common Kaikki part-of-speech codes; anything not
# listed is passed through verbatim.
_POS_LABEL = {
    "adj": "adjective",
    "adv": "adverb",
    "name": "proper noun",
    "num": "numeral",
    "pron": "pronoun",
    "det": "determiner",
    "prep": "preposition",
    "conj": "conjunction",
    "intj": "interjection",
}

# language -> {form_lower: [(pos_code, description), ...]}
_cache = {}


def _load(language):
    if language in _cache:
        return _cache[language]
    index = {}
    path = _DIR / f"{language}.jsonl"
    if path.exists():
        with open(path, encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if not line:
                    continue
                try:
                    rec = json.loads(line)
                except json.JSONDecodeError:
                    continue
                form = rec.get("form")
                if not form:
                    continue
                out = []
                for a in rec.get("analyses", []):
                    pos = a.get("pos")
                    tags = a.get("tags")
                    lemma = a.get("lemma")
                    bits = []
                    if pos:
                        bits.append(_POS_LABEL.get(pos, pos))
                    if tags:
                        bits.append(tags)
                    text = ", ".join(bits)
                    if lemma and lemma.lower() != form:
                        text += f' (of "{lemma}")'
                    pair = (pos, text)
                    if pair not in out:
                        out.append(pair)
                index[form] = out
    _cache[language] = index
    return index


def describe_form(word, language):
    """`[(pos_code, description), ...]` for the exact `word` in
    `language`, or `[]` if it isn't in the table. Never raises, never
    touches the network."""
    return _load(language).get(word.lower(), [])

File: backend_java/src/main/java/falcon/InflectionLookup.java
package falcon;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Pure-local grammatical analysis of an exact inflected word form, read from
 * data/inflection/&lt;lang&gt;.jsonl (mirrors backend/inflection_lookup.py).
 */
public final class InflectionLookup {
    private InflectionLookup() {}

    private static final Path DIR = Env.path("data", "inflection");

    private static final Map<String, String> POS_LABEL = Map.of(
            "adj", "adjective",
            "adv", "adverb",
            "name", "proper noun",
            "num", "numeral",
            "pron", "pronoun",
            "det", "determiner",
            "prep", "preposition",
            "conj", "conjunction",
            "intj", "interjection");

    /** (pos_code, description) pair. */
    public record Analysis(String pos, String description) {}

    private static final Map<String, Map<String, List<Analysis>>> CACHE = new ConcurrentHashMap<>();

    private static Map<String, List<Analysis>> load(String language) {
        return CACHE.computeIfAbsent(language, InflectionLookup::build);
    }

    private static Map<String, List<Analysis>> build(String language) {
        Map<String, List<Analysis>> index = new HashMap<>();
        Path path = DIR.resolve(language + ".jsonl");
        if (!Files.exists(path)) return index;
        try (BufferedReader r = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
            String line;
            while ((line = r.readLine()) != null) {
                line = Py.strip(line);
                if (line.isEmpty()) continue;
                Map<String, Object> rec;
                try {
                    rec = Json.parseObject(line);
                } catch (IllegalArgumentException e) {
                    continue;
                }
                String form = Json.str(rec, "form", null);
                if (form == null || form.isEmpty()) continue;
                List<Analysis> out = new ArrayList<>();
                for (Object a : Json.listOrEmpty(rec.get("analyses"))) {
                    String pos = Json.str(a, "pos", null);
                    String tags = Json.str(a, "tags", null);
                    String lemma = Json.str(a, "lemma", null);
                    List<String> bits = new ArrayList<>();
                    if (pos != null && !pos.isEmpty()) bits.add(POS_LABEL.getOrDefault(pos, pos));
                    if (tags != null && !tags.isEmpty()) bits.add(tags);
                    String text = String.join(", ", bits);
                    if (lemma != null && !lemma.isEmpty() && !lemma.toLowerCase(Locale.ROOT).equals(form)) {
                        text += " (of \"" + lemma + "\")";
                    }
                    Analysis pair = new Analysis(pos, text);
                    if (!out.contains(pair)) out.add(pair);
                }
                index.put(form, out);
            }
        } catch (IOException e) {
            Env.log("inflection_lookup: cannot read " + path + ": " + e.getMessage());
        }
        return index;
    }

    /** Analyses of the exact {@code word}, or an empty list. Never throws. */
    public static List<Analysis> describeForm(String word, String language) {
        return load(language).getOrDefault(word.toLowerCase(Locale.ROOT), List.of());
    }
}
